### PR TITLE
Strong Password implementation

### LINKED ISSUE #1461: Implement the password_strength password gem
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**

A soft-ish spot that we should revisit according to best practices are our practices around passwords. This isn't terribly urgent since much of our userbase is using oauth, but it couldn't hurt to crank this up for the stragglers.

* **What feature or behavior is this required for?**

~ le security ~

* **How could we solve this issue? (Not knowing is okay!)**

We want to add the [Password Strength](https://github.com/fnando/password_strength) gem to our validation step.

* **Anything else?**

Original Brainstorming: 

<details>
Via our security audit.

These lists may come in handy: https://github.com/danielmiessler/SecLists/tree/master/Passwords\

I'm interested in two areas in particular: preventing commonly used passwords from being used, and cranking up the length. The length thing is straightforward enough, but I am interested in whether or not there's an out of the box ruby util we can check against. Worth investigating!

</details>

### PR DESCRIPTION
**Pull request by lomky**
Pushing forward on Strong Passwords! Opening a new PR with the code plucked and separated from some editor linting that was plaguing #2067! Many kudos to @ccedacero for the work I'm building on here.

This pull request makes the following changes:

  -  Added strong password gem
  -  Added password validation for common dictionary words, 40 entropy, and 8 chars length
  -  Changed seed file to reflect password strength. 
  - Adds tests

It relates to the following issue #s: 
  - Fixes #1461
  - Replaces #2067 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -142,6 +142,10 @@ en:
         sign_in: Sign in
       user:
         signed_in: ''
+  errors:
+    messages:
+      password:
+        password_strength: Passwords need to be stronger than that. Try a longer or more complicated password please
```
Comment (colinxfleming) [inline config/locales/en.yml:148]: bummer that the gem doesn't have a sensible default here!

File: app/models/user.rb [hunk 2]
```diff
@@ -82,12 +82,18 @@ class User
...
   validates :role, presence: true
-  validate :secure_password
+  validate :secure_password, if: :updating_password?
+  # i18n-tasks-use t('errors.messages.password.password_strength')
+  validates :password, password_strength: {use_dictionary: true}, if: :updating_password?
+
 
   TIME_BEFORE_DISABLED_BY_FUND = 9.months
 
+  def updating_password?
```
Comment (colinxfleming) [inline app/models/user.rb:92]: I think there's a nicer pg magic method we'll be able to use when we switch over; we'll have to check

**Review by colinxfleming — APPROVED** [review]
This scans to me. Since this is pretty user affecting, we should telegraph that we're doing it soonish, and then send a reminder when we ship it to production. (I'll write that blurb up)

**Comment by colinxfleming** [comment]
Thanks for getting this over the finish line @lomky ! Didn't smash merge since I figured you wanted to give @xmunoz a chance to weigh in, but please do smash the button when you feel good.

File: Gemfile [hunk 3]
```diff
@@ -37,6 +37,9 @@ gem 'paper_trail-globalid'
 gem 'devise', '~> 4.7'
 gem 'omniauth-google-oauth2', '~> 0.8.0'
 
+# Strong Password for user password validation for folks not on oauth
+gem 'strong_password', '~> 0.0.9'
```
Comment (xmunoz) [inline Gemfile:41]: I tried the look at the source for this version, and found that the dev actually hasn't published the tag yet: https://github.com/bdmac/strong_password/issues/38

This is not a problem with your PR (0.0.9 is the latest version published on rubygems), just something to be aware of.

File: README.md [hunk 4]
```diff
@@ -4,9 +4,10 @@
 [![](https://images.microbadger.com/badges/image/colinxfleming/dcaf_case_management.svg)](https://microbadger.com/images/colinxfleming/dcaf_case_management "Get your own image badge on microbadger.com")
 
 [A deployed demo version of what's in the main branch is at: https://sandbox.dcabortionfund.org/](https://sandbox.dcabortionfund.org/)  
-User: test@example.com, Password: P4ssword
+User: test@example.com, Password: AbortionsAreAHumanRight1
```
Comment (xmunoz) [inline README.md:7]: <3

File: app/models/user.rb [hunk 5]
```diff
@@ -82,12 +82,18 @@ class User
   # email presence validated through Devise
   validates :name, presence: true
   validates :role, presence: true
-  validate :secure_password
+  validate :secure_password, if: :updating_password?
+  # i18n-tasks-use t('errors.messages.password.password_strength')
```
Comment (xmunoz) [inline app/models/user.rb:86]: Why is this commented out?

File: app/models/user.rb [hunk 6]
```diff
@@ -82,12 +82,18 @@ class User
...
+  validates :password, password_strength: {use_dictionary: true}, if: :updating_password?
```
Comment (xmunoz) [inline app/models/user.rb:87]: Should we also add `min_word_length`? Docs say that the default is 4, I recommend setting it to 6 or 8.

File: config/locales/en.yml (hunk 1 above)
Comment (xmunoz) [reply config/locales/en.yml:148]: It actually looks like it does: https://github.com/bdmac/strong_password/blob/75a2c79df5c54adb97b79f1c630a3162370e022e/README.md#with-railsactivemodel-validations

> The default validation message is "is too weak". Rails will display a field of :password having too weak of a password with full_error_messages as "Password is too weak"

File: db/seeds.rb [hunk 7]
```diff
@@ -8,20 +8,25 @@
 Event.destroy_all
 Config.destroy_all
 
+# Create google SSO user
```
Comment (xmunoz) [inline db/seeds.rb:11]: Why is this commented out? Is this documentation? If so, can you move it to the docs?

File: config/locales/en.yml (hunk 1 above)
Comment (xmunoz) [inline config/locales/en.yml:148]:
```suggestion
        password_strength: Passwords need to be stronger than that. Try a longer or more complicated password please.
```

File: app/models/user.rb (hunk 5 above)
Comment (lomky) [reply app/models/user.rb:86]: it tells our i18n linter that we do use the value

File: test/models/user_test.rb [hunk 8]
```diff
@@ -28,6 +28,31 @@ class UserTest < ActiveSupport::TestCase
                    'include DCAF and password.',
                    @user.errors.messages[:password].first
     end
+
+    it 'should require a strong password' do
+      @user.password = 'Hello#2020'
+      @user.password_confirmation = 'Hello#2020'
+      assert_not @user.valid?
+      assert_equal 'Passwords need to be stronger than that. ' \
+                   'Try a longer or more complicated password please',
```
Comment (lomky) [inline test/models/user_test.rb:37]:
```suggestion
                   'Try a longer or more complicated password please.',
```

**Review by xmunoz — APPROVED** [review]

**Comment by colinxfleming** [comment]
I still owe the team comms around this before we deploy, but I think we can merge it now.

**Comment by lomky** [comment]
@all-contributors please add @ccedacero for code! 💚

**Pull request merged**